src/ServiceControl/Infrastructure/MetadataExtensions.cs

```diff
@@ -0,0 +1,41 @@
+﻿namespace ServiceControl
+{
+    using System;
+    using System.Collections.Generic;
+
+    static class MetadataExtensions
+    {
+        public static T MaybeGet<T>(this IDictionary<string, object> metadata, string key)
+        {
+            object foundValue;
+            if (metadata.TryGetValue(key, out foundValue))
+            {
+                return (T)foundValue;
+            }
+            return default(T);
+        }
+
+        public static string MaybeGetAsString(this IDictionary<string, object> metadata, string key)
+        {
+            object foundValue;
+            if (metadata.TryGetValue(key, out foundValue))
+            {
+                return foundValue?.ToString();
+            }
+
+            return null;
+        }
+
+        public static DateTime? AsMaybeDateTime(this string s)
+        {
+            if (s == null) return null;
+            DateTime dt;
+            if (DateTime.TryParse(s, out dt))
```

Comment:
Wouldn't we want to specify the datetime format we want to see here as well as the datetime style and the invariant culture?

Reply:
Possibly.

The previous code was just `DateTime.Parse` but it was immediately after a `ToString()`. We might be able to just turn it into a `GetOrDefault<DateTime>()`. I'm not sure why it was being round-tripped through `System.String`.

Reply:
Maybe then doing TryParse should work. I have not enough knowledge here to say it should work or it won't